src/mococrw/dilithium.h

```diff
@@ -0,0 +1,347 @@
+ */
+
+/**
+ * This class is a replacement of EVP_PKEY as long as dilithium is not supported by the openssl
+ * version used by MoCOCrW. It stores the key as std::vector<uint8_t> and the key type.
+ * Addittionally some convenient functions are added.
+ */
+// Rename DilithiumKeyImplementation
+class DilithiumKeyImpl
+{
+public:
+    DilithiumKeyImpl(const std::vector<uint8_t> key_data, AsymmetricKey::KeyTypes key_type)
+            : _keyType(key_type), _key_data{std::move(key_data)}
+    {
+        if (!hasValidKeySize()) {
+            throw MoCOCrWException("Key data does not match the expected size.");
+        }
+    }
+
+    static std::shared_ptr<DilithiumKeyImpl> parseAsn1PublicKey(
+            const std::vector<uint8_t> &x509PubKey);
+    static std::shared_ptr<DilithiumKeyImpl> parseAsn1PrivateKey(
+            const std::vector<uint8_t> &pkcs8PrivKey);
+    /**
+     * @returns The key type
+     */
+    AsymmetricKey::KeyTypes getType() const { return _keyType; };
+
+    /**
+     * @returns The size of the raw key data
+     */
+    uint getKeySize() const noexcept { return _key_data.size(); }
+
+    /**
+     * Returns the raw key data
+     *
+     * @returns The raw key data
+     */
+    const std::vector<uint8_t> getKeyData() const { return _key_data; };
+
+    /**
+     * Returns the public key. If a private key is stored the public key is extracted and returned
+     *
+     * @return The public key
+     * @throws If the stored key size doesn't match the size of a private or public key
+     * @throws If the extraction functin to get the public key from a private key returns an error
+     * @throws If the key type is not supported
+     */
+    DilithiumKeyImpl getPublicKey() const;
+
+    /**
+     * @brief Checks if the key has a valid key size
+     *
+     * @return true if the key matches either the length of the private or public key of the given
+     * key type
+     * @return false otherwise
+     */
+    bool hasValidKeySize() const;
+
+    /**
+     * Checks if the key size matches the size of a private key. Does not do any validity checks.
+     * The key size is dependent on the selected key type (Dilithium3 or Dilithium5).
+     *
+     * @return True if the key size matches the size of private keys.
+     * @throws If the key size neither matches a private or public key size.
+     * @throws if the key type is not supported
+     */
+    bool isPrivateKey() const;
+
+    inline bool operator==(DilithiumKeyImpl &key) { return _key_data == key.getKeyData(); }
+
+private:
+    const AsymmetricKey::KeyTypes _keyType;
+    const std::vector<uint8_t> _key_data;
+};
+
+/**
+ * @brief Dilithium representation of AsymmetricKey
+ *
+ * @see AsymmetricKey
+ *
+ * This shall be replaced by AsymmetricKey once Dilithium is supported by the openssl version
+ * used by MoCOCrW.
+ * The interface is similar to the one of AsymmetricKey.
+ *
+ * For documentation see AsymmetricKey
+ */
+class DilithiumAsymmetricKey
```

Comment:
The idea was to duplicate the interface of `key.h`. Surely it would be possible to add this functionality into `DilithiumKeyImpl`, however the rational was to be able to replace `dilithium.h` with `key.h` once openssl supports dilithium.
But I am open for improvements. :)
Same for the other classes.
I could come up with a wrapping interface for the classes in `key.h` without doing an API break.

Reply:
hm. The `AsymmetricKey` class is a bit weird because its more or less some plain encapsulation of the OpenSSL things and then just serve the Public and the Keypair classes. The customer code only interacts with the latter two and just uses some of the definitions from `AsymmetricKey`. Maybe we can put the definitions already into AsymmetricKey and otherwise omit this whole class here? I'm not sure how valuable the contribution of another indirection is here.

Reply:
as discussed I leave the structure as it is, to simplify the adaptions once openssl supports dilithium.